### ISSUE TITLE
One style to rule them all

### ISSUE THREAD

**Issue by chrisdone** [chrisdone/hindent#207]
This issue is to track the migration of hindent from handling multiple styles to handling just one:

https://github.com/chrisdone/hindent/blob/master/MIGRATION.md

> hindent will now only support one style, based on the Johan Tibell
> style.
> 
> * The API will move away from supporting different styles.
> * The command-line interface will remove the `--style` flag, ignoring
>   it silently.
> * Issues related to different styles will be closed.
> * The test suite will be restricted to the one style.

So far there's a measurable speed improvement of 20x:

```
benchmarking Main/BigDeclarations/HIndent.reformat: "johan-tibell"
time                 364.7 ms   (345.0 ms .. 387.9 ms)
                     1.000 R²   (0.998 R² .. 1.000 R²)
mean                 354.4 ms   (345.4 ms .. 359.1 ms)
std dev              7.777 ms   (67.99 as .. 8.176 ms)
variance introduced by outliers: 19% (moderately inflated)
```

after

```
benchmarking Main/BigDeclarations/HIndent.reformat
time                 17.56 ms   (17.09 ms .. 18.05 ms)
                     0.996 R²   (0.992 R² .. 0.998 R²)
mean                 17.89 ms   (17.60 ms .. 18.20 ms)
std dev              748.3 μs   (587.7 μs .. 1.030 ms)
variance introduced by outliers: 16% (moderately inflated)
```

**Comment by gibiansky** [comment]
I don't understand why removing styles would result in a speed improvement... Do you know why this is?

Besides that, while I'm pretty down with the idea of having just one style, removing all other styles is not how I would go about it.  My intuition (although I have yet to test this fully) is that the `johan-tibell` style does not reflect how I write Haskell and isn't flexible enough, so I'm planning on maintaining `hindent-4.*` at [`gibiansky/hindent`](https://github.com/gibiansky/hindent) for the foreseeable future. Not sure how much maintenance vs just leaving it there so that I can continue to use my own style, but I'm happy to accept issues and PRs related to non-johan-tibell styles, so feel free to redirect people there if necessary.

**Comment by chrisdone** [comment]
> I don't understand why removing styles would result in a speed improvement... Do you know why this is?

I think the Typeable-based dispatching slowed things down significantly. 

> the johan-tibell style does not reflect how I write Haskell and isn't flexible enough, so I'm planning on maintaining hindent-4.* at gibiansky/hindent for the foreseeable future. Not sure how much maintenance vs just leaving it there so that I can continue to use my own style, but I'm happy to accept issues and PRs related to non-johan-tibell styles, so feel free to redirect people there if necessary.

Cool. I think this is where we as a community can converge on something. If the one style isn't flexible enough, we add more to it, or add options. But in the end it's just one style. I dropped my own radically different style in favour of the tibbe one to cooperate with other people on my teams. I found that with automation, I care less about the style, because it's at least consistent. It doesn't work as well with SHM, but I can update SHM to work better with it.

**Comment by gibiansky** [comment]
> Cool. I think this is where we as a community can converge on something. If the one style isn't flexible enough, we add more to it, or add options.

I agree! Sadly, right now I don't have time to port all the things I think are critical from the `gibiansky` style to the `johan-tibell` style; in addition, if it's a single "community style", it's hard to say *what* should be ported over. While I like the idea of a single, community style, I would prefer a path in which we add a new style `community` which is sourced by *asking* the wider community what they consider to be the canonical style.

But, as you can imagine, that has all sorts of complications and requires quite a bit of development. Right now I don't have the bandwidth to do any of that, and in the meantime am more-or-less happy with `hindent` for my own personal use, so will at least temporarily be maintaining a fork so that I can continue Getting Shit Done (TM) on other ongoing projects. Hopefully this is temporary; I'm not super interested in having TWO `hindent` copies floating around, so hopefully at some point the now-default `johan-tibell` style will evolve into something that is as usable as `gibiansky` is for me, at which point I can happily drop the fork. (For me this is a case of if-it-aint-broke-dont-fix-it: `hindent` is fast enough for me, so not sure what we gain by dropping styles.)

That said, IMHO, if you would like to create a single, agreed-upon style, it needs to be clear that the style is a) sourced from wider community preferences and b) has a procedure for modifying it.

**Comment by alanz** [comment]
I think the `johan-tibell` style is the closest you will get to a community-endorsed standard, anything else will just trigger major bikeshedding.

**Comment by cocreature** [comment]
I’m mostly fine with the `johan-tibell` style but personally I would like an option to control the indentation width. There are a lot of projects that use 2 spaces and it’d be a shame if I had to stop using hindent on all of them.

**Comment by chrisdone** [comment]
All good points I agree with.

**Comment by danielwaterworth** [comment]
Having looked at this more closely, I think the speed increase came from making the `m` in `Pretty` concrete rather than reducing `Typeable` checks.

**Comment by chrisdone** [comment]
This is done.

**Comment by chrisdone** [comment]
The benchmarks are stored in https://github.com/chrisdone/hindent/blob/master/BENCHMARKS.md

And the tests are stored in https://github.com/chrisdone/hindent/blob/master/TESTS.md

I made a trivial markdown parser so that both suites parse those files into criterion/hspec trees, which has the nice benefit that people can view the `TESTS.md` on GitHub as a kind of style spec.

**Comment by chrisdone** [comment]
@gibiansky @danielwaterworth @ocharles 

I just introduced an optimization I've wanted to try out for a while which was short-circuiting the printer when testing to see if a line-break is needed (by either exceeding the line limit or going to a newline). I implemented that via the existing `MaybeT` type. Normally `fitsOnOneLine` (the point of the exponential explosion) will run a `Printer a` and return whether it exceeded the bounds. This happens recursively. Whether or not it returned `True` or `False` it would go ahead and print the complete tree out even it was to be discarded.

Now, as soon as it hits the column limit or a newline, it immediately returns "no". The performance benefits are substantial:

From
    
    benchmarking Large inputs/Bunch of declarations
    time                 81.14 ms   (78.49 ms .. 83.51 ms)
                         0.998 R²   (0.996 R² .. 1.000 R²)
    mean                 80.68 ms   (79.76 ms .. 81.61 ms)
    std dev              1.547 ms   (1.139 ms .. 2.002 ms)
    
    benchmarking Complex inputs/Quasi-quotes with nested lets and operators
    time                 1.020 s    (990.7 ms .. 1.046 s)
                         1.000 R²   (1.000 R² .. 1.000 R²)
    mean                 1.034 s    (1.026 s .. 1.040 s)
    std dev              9.228 ms   (0.0 s .. 10.32 ms)
    variance introduced by outliers: 19% (moderately inflated)
    
to
    
    benchmarking Large inputs/Bunch of declarations
    time                 6.272 ms   (5.823 ms .. 6.745 ms)
                         0.982 R²   (0.970 R² .. 0.997 R²)
    mean                 5.803 ms   (5.707 ms .. 5.962 ms)
    std dev              360.2 μs   (238.6 μs .. 604.3 μs)
    variance introduced by outliers: 35% (moderately inflated)
    
    benchmarking Complex inputs/Quasi-quotes with nested lets and operators
    time                 2.549 ms   (2.460 ms .. 2.627 ms)
                         0.993 R²   (0.990 R² .. 0.996 R²)
    mean                 2.685 ms   (2.632 ms .. 2.757 ms)
    std dev              193.2 μs   (163.5 μs .. 229.2 μs)
    variance introduced by outliers: 50% (severely inflated)

**Comment by ocharles** [comment]
This is really good to hear - I'm not sure if I'd mentioned this to you or not in the past, but I'd also thought about doing something similar! Anywhere I can try this out?

**Comment by chrisdone** [comment]
@ocharles I think we discussed it on IRC at least once. This is currently on master. 😄

**Comment by filib** [comment]
This is awesome. Massive kudos.

**Comment by chrisdone** [comment]
@ocharles Forgot to mention there's a demonstration site: http://chrisdone.com/hindent

Rewriting the comment processing brought that 6ms up to 12ms. But then I added some more short-circuiting between passes, so the same input without any comments is 2ms. This may be bordering on micro-optimization, but I anticipate it'll save on large files.

I also switched to the bytestring-builder at some point in this rather than the text builder.

Bumped on hackage as 5.1.0.
